Add Cython implementation for VOTable binary converters

stvoutsin opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description

This pull request attempts to improve the performance of VOTable binary parsing by implementing optimized Cython converters.

Initial benchmarks have shown performance improvements of the order:
- 53% speedup on numeric-heavy tables 
- 43-47% speedup on mixed data type tables
- 30-35% speedup on string-heavy tables
- Consistent improvements across BINARY and BINARY2 formats

**Problem:**
We’re seeing relatively slow performance when parsing large VOTables using the BINARY2 serialization format. 
 Profiling results from py-spy show a significant portion of time is spent in astropy.io.votable.converters.binparse. Here’s an excerpt of the profiler output:

```
py-spy top -- python test.py
Collecting samples from 'python test.py' (python v3.12.7)
Total Samples 1600
GIL: 100.00%, Active: 100.00%, Threads: 1
  %Own   %Total  OwnTime  TotalTime  Function (filename)                                                                                                                                                           
 54.00%  57.00%    7.69s     8.26s   binparse (astropy/io/votable/converters.py)
 24.00% 100.00%    3.68s    14.66s   parsebinary (astropy/io/votable/tree.py)
  4.00%   4.00%   0.940s    0.940s   bitarray_to_bool (astropy/io/votable/converters.py)
  8.00%   8.00%   0.890s    0.910s   setitem (numpy/ma/core.py)
  0.00%   0.00%   0.470s    0.670s   getbinary_data_stream (astropy/io/votable/tree.py)
  2.00%   2.00%   0.450s    0.450s   careful_read (astropy/io/votable/tree.py)
  0.00%   0.00%   0.430s    0.430s   new (numpy/_core/records.py)
  
```
As shown, the hot path is dominated by the binparse function in converters.py.


**Solution:**
- Added "Fast" Cython converters for all numeric types (`double`, `float`, `int`, `long`, `short`, `unsignedByte`, `boolean`, `bit`)

**Compatibility:**
Should be 100% backward compatible since the API has not been modified

**Testing:**
Added comprehensive benchmarks to astropy-benchmarks showing consistent 30-50% performance improvements across different table types and data patterns. https://github.com/astropy/astropy-benchmarks/pull/141

Astropy:main

```
[ 0.00%] · For astropy commit 8599d499 <main>:
[ 0.00%] ·· Benchmarking virtualenv-py3.11
[ 3.85%] ··· votable.TimeVOTableBooleanFields.time_booleans_binary                                                                                                                                         2.19±0s
[ 7.69%] ··· votable.TimeVOTableBooleanFields.time_booleans_binary2                                                                                                                                        3.16±0s
[11.54%] ··· votable.TimeVOTableLongStrings.time_long_strings_binary                                                                                                                                       2.45±0s
[15.38%] ··· votable.TimeVOTableLongStrings.time_long_strings_binary2                                                                                                                                      3.31±0s
[19.23%] ··· votable.TimeVOTableMixed.time_mixed_binary                                                                                                                                                    4.37±0s
[23.08%] ··· votable.TimeVOTableMixed.time_mixed_binary2                                                                                                                                                   5.77±0s
[26.92%] ··· votable.TimeVOTableNumeric.time_numeric_binary                                                                                                                                                3.37±0s
[30.77%] ··· votable.TimeVOTableNumeric.time_numeric_binary2                                                                                                                                               4.24±0s
[34.62%] ··· votable.TimeVOTableShortStrings.time_short_strings_binary                                                                                                                                     2.98±0s
[38.46%] ··· votable.TimeVOTableShortStrings.time_short_strings_binary2                                                                                                                                    3.82±0s
[42.31%] ··· votable.TimeVOTableSmallOverhead.time_small_binary                                                                                                                                           10.5±0ms
[46.15%] ··· votable.TimeVOTableSmallOverhead.time_small_binary2                                                                                                                                          13.9±0ms
[50.00%] ··· votable.TimeVOTableStringIntensive.time_string_intensive_binary2                                                                                                                              4.42±0s
```
u/stvoutsin/binary2-cython 

```
[ 0.00%] · For astropy commit 82cd430e <u/stvoutsin/binary2-cython>:
[ 0.00%] ·· Benchmarking virtualenv-py3.11
[ 3.85%] ··· votable.TimeVOTableBooleanFields.time_booleans_binary                                                                                                                                         1.82±0s
[ 7.69%] ··· votable.TimeVOTableBooleanFields.time_booleans_binary2                                                                                                                                        2.38±0s
[11.54%] ··· votable.TimeVOTableLongStrings.time_long_strings_binary                                                                                                                                       1.63±0s
[15.38%] ··· votable.TimeVOTableLongStrings.time_long_strings_binary2                                                                                                                                      2.20±0s
[19.23%] ··· votable.TimeVOTableMixed.time_mixed_binary                                                                                                                                                    2.51±0s
[23.08%] ··· votable.TimeVOTableMixed.time_mixed_binary2                                                                                                                                                   3.05±0s
[26.92%] ··· votable.TimeVOTableNumeric.time_numeric_binary                                                                                                                                                1.57±0s
[30.77%] ··· votable.TimeVOTableNumeric.time_numeric_binary2                                                                                                                                               2.01±0s
[34.62%] ··· votable.TimeVOTableShortStrings.time_short_strings_binary                                                                                                                                     1.96±0s
[38.46%] ··· votable.TimeVOTableShortStrings.time_short_strings_binary2                                                                                                                                    2.45±0s
[42.31%] ··· votable.TimeVOTableSmallOverhead.time_small_binary                                                                                                                                           6.12±0ms
[46.15%] ··· votable.TimeVOTableSmallOverhead.time_small_binary2                                                                                                                                          8.35±0ms
[50.00%] ··· votable.TimeVOTableStringIntensive.time_string_intensive_binary2                                                                                                                              3.26±0s
```
For better visualization:

| Benchmark | Before | After | Improvement |
|-----------|--------|-------|-------------|
| Numeric (BINARY) | 3.37s | 1.57s | 53% faster |
| Numeric (BINARY2) | 4.24s | 2.01s | 53% faster |
| Mixed (BINARY) | 4.37s | 2.51s | 43% faster |
| Mixed (BINARY2) | 5.77s | 3.05s | 47% faster |
| Long Strings (BINARY) | 2.45s | 1.63s | 33% faster |
| Long Strings (BINARY2) | 3.31s | 2.20s | 33% faster |
| Short Strings (BINARY) | 2.98s | 1.96s | 34% faster |
| Short Strings (BINARY2) | 3.82s | 2.45s | 36% faster |
| Boolean Fields (BINARY) | 2.19s | 1.82s | 17% faster |
| Boolean Fields (BINARY2) | 3.16s | 2.38s | 25% faster |
| Small Overhead (BINARY) | 10.5ms | 6.12ms | 42% faster |
| Small Overhead (BINARY2) | 13.9ms | 8.35ms | 40% faster |
| String Intensive (BINARY2) | 4.42s | 3.26s | 26% faster |


I've also tested parsing outside of the astropy-benchmark PR with a VOTable with 50000 rows and aprox. 1000 columns (~50 million cells).
With this table the parsing took 30 seconds, down from 65 seconds in the version currently on main.

<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->


<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #18442
<!-- Optional opt-out -->

Any thoughts on this approach? Are there alternatives that I haven't considered? Perhaps someone with more Cython experience can let me know if I've made any obvious mistakes or if there are better way of doing any of this.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

stvoutsin committed:
Add Cython implementation for VOTable binary converters

pllim commented on astropy/io/votable/converters.py:
```diff
@@ -842,6 +828,12 @@ class Double(FloatingPoint):
 
     format = "f8"
 
+    def binparse(self, read):
```
Seems a little repetitive across the board. Looks like the only difference is a number. Can this method be inherited but access some `self._expected_len` (name negotiable) that is set by the subclass?

stvoutsin replied:
Great suggestion, looking back at it now I can see there is a lot of duplication. I've moved things up to the parent Numeric class, let me know if that looks good. (Or if you see any adjustments needing to be made)

pllim commented on docs/changes/io.votable/18454.perf.rst:
Once we have the benchmark down, should maybe put some ballpark values here.

pllim commented:
Hmm looks like @eerovaher proposed an alternative at https://github.com/astropy/astropy/pull/18455 . Are you interested to have a look, @stvoutsin ? Thanks, all!

pllim commented:
https://github.com/astropy/astropy-benchmarks/pull/141 is merged so theoretically next push would trigger a new benchmark run with the new relevant benchmarks. 🤞

stvoutsin committed:
Fix buffer handling and null detection to match original converter behavior

stvoutsin committed:
Reduce duplication of binparse methods for Numeric types

stvoutsin committed:
Move null handling to Cython parser implementations to address performance regression

stvoutsin commented:
> significantly changed"

Apologies I think i introduced a regression while trying to fix an issue that caused failed tests in the CI. 
I've changed this approach so let's see if this run gives better results. 
At least locally, I get:

```
[ 3.85%] ··· votable.TimeVOTableBooleanFields.time_booleans_binary                                                                                                                                         1.40±0s
[ 7.69%] ··· votable.TimeVOTableBooleanFields.time_booleans_binary2                                                                                                                                        1.94±0s
[11.54%] ··· votable.TimeVOTableLongStrings.time_long_strings_binary                                                                                                                                       1.26±0s
[15.38%] ··· votable.TimeVOTableLongStrings.time_long_strings_binary2                                                                                                                                      1.75±0s
[19.23%] ··· votable.TimeVOTableMixed.time_mixed_binary                                                                                                                                                    2.01±0s
[23.08%] ··· votable.TimeVOTableMixed.time_mixed_binary2                                                                                                                                                   2.60±0s
[26.92%] ··· votable.TimeVOTableNumeric.time_numeric_binary                                                                                                                                                1.27±0s
[30.77%] ··· votable.TimeVOTableNumeric.time_numeric_binary2                                                                                                                                               1.80±0s
[34.62%] ··· votable.TimeVOTableShortStrings.time_short_strings_binary                                                                                                                                     1.53±0s
[38.46%] ··· votable.TimeVOTableShortStrings.time_short_strings_binary2                                                                                                                                    2.07±0s
[42.31%] ··· votable.TimeVOTableSmallOverhead.time_small_binary                                                                                                                                           5.33±0ms
[46.15%] ··· votable.TimeVOTableSmallOverhead.time_small_binary2                                                                                                                                          7.81±0ms
[50.00%] ··· votable.TimeVOTableStringIntensive.time_string_intensive_binary2                                                                                                                              2.80±0s
```

I will also test with increasing the row sizes for the sample data to see how the runtime difference between main and this branch scales.

stvoutsin committed:
Update docs/change entry with some notes on benchmark results

pllim commented:
Benchmark run looking good. Thanks!

| Change   | Before [db7a3e2e]    | After [30bfc4a5]    |   Ratio | Benchmark (Parameter)                                     |
|----------|----------------------|---------------------|---------|-----------------------------------------------------------|
| -        | 2.82±0s              | 2.00±0s             |    0.71 | votable.TimeVOTableShortStrings.time_short_strings_binary |
| -        | 3.09±0s              | 2.03±0.01s          |    0.66 | votable.TimeVOTableLongStrings.time_long_strings_binary2  |
| -        | 13.6±0.09ms          | 8.81±0.04ms         |    0.65 | votable.TimeVOTableSmallOverhead.time_small_binary2       |
| -        | 10.4±0.1ms           | 6.27±0.03ms         |    0.61 | votable.TimeVOTableSmallOverhead.time_small_binary        |
| -        | 5.60±0s              | 3.35±0.01s          |    0.6  | votable.TimeVOTableMixed.time_mixed_binary2               |
| -        | 4.43±0.01s           | 2.26±0s             |    0.51 | votable.TimeVOTableNumeric.time_numeric_binary2           |
| -        | 3.51±0.01s           | 1.69±0s             |    0.48 | votable.TimeVOTableNumeric.time_numeric_binary            |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.

stvoutsin commented:
Any thoughts on possibly moving this forward?

stvoutsin committed:
Revert bitarray converters to use numpy instead of Cython

bsipocz commented:
@stvoutsin - could you please rebase this so we get fresh CI status? I'll try to review before feature freeze.

[never mind, it looks like we can now do that from the UI]

bsipocz commented:
CI failures, all unrelated and seem to be present on master: 
   - weekly cron -> unrelated broken links
   - py3.14 and devdeps -> somethink with samp? and maybe uncertainty? doesn't look related
   - simulated arch - failures in table

pllim commented:
To be specific:

* weekly cron -> #19637 (a rebase would do it but also not critical)
* py3.14 and devdeps -> #19621 (it is numpy-dev)
* simulated arch -> #19632

pllim commented:
@bsipocz are we merging this in for v8.0.0 ? Looks good to you now?

bsipocz commented on astropy/io/votable/src/fast_converters.pyx:
```diff
@@ -0,0 +1,266 @@
+        value = True
+    elif byte_val == ord(b'F') or byte_val == ord(b'f') or byte_val == ord(b'0'):
+        value = False
+    elif byte_val == 0 or byte_val == ord(b' ') or byte_val == ord(b'?'):
+        value = False
+        is_null = True
+    else:
+        value = False
+        is_null = True
```
why do we need this elif? I mean I see that in the non cython version we have the mapping for these, but here maybe it's not needed?

stvoutsin replied:
Oh yea that looks unnecessary, thanks for spotting. I'll remove in a quick follow-up PR.

pllim replied:
Seems simple enough that I already opened https://github.com/astropy/astropy/pull/19664 so we don't forget about this before RC. 😅  Please review, thanks!

bsipocz reviewed (approved):
OK, I think this is good to go. I tried to review the changes, especially where it's not just the addition but the removal of existing python lines; but didn't run any additional round tripping tests, but knowing that @stvoutsin will be a user of these I trust any problems will eventually smoked out.

Minor comment is not a blocker for merging.

bsipocz commented:
Thanks @stvoutsin again and also thanks for being patient with this.

stvoutsin commented:
> OK, I think this is good to go. I tried to review the changes, especially where it's not just the addition but the removal of existing python lines; but didn't run any additional round tripping tests, but knowing that @stvoutsin will be a user of these I trust any problems will eventually smoked out.
> 
> Minor comment is not a blocker for merging.

Cool thanks for reviewing, and indeed we do rely on this so we're happy to take up any work needed to support this!

bsipocz commented:
The main concern was to see easier ways to spot that there aren't any regressions/changes in the actual data values. But then we decided that you are the best person to test/benchmark that so here we're 😄

pllim commented:
Please also review https://github.com/astropy/astropy/pull/19686 to ignore the C file that this PR generates. Thanks!
